Add C bindings

laplus-sadness opened the pull request:
Hello. I discovered this project through [TWIR](https://this-week-in-rust.org/blog/2023/07/05/this-week-in-rust-502/) and wanted to contribute, so I did some work on the #209 issue.

Problem is I don't know if you would like my implementation, so I didn't write a readme or any tests, but I tested on my computer and worked correctly.

Some notes about things I thought while working on this:
- Setting the `output` pointer from `inline_to` to null in the case of an error wouldn't work out because, if the array is allocated, the user would need to make a copy of it before running the function or else they would lose the address.
- I like [FatFs](http://elm-chan.org/fsw/ff/00index_e.html) return types, an enum, instead of -1 or -2 for errors. So the code would be:
```c
int res = inline_to(...);
if (res != CSS_RESULT_OK) { // or !res
    // clean
}
``` 
- I couldn't discover how to use `[[no_discard]]` with cbindgen, so the header is without this parameter.
- I didn't add `rust-version = "1.60"` on Cargo.toml, don't know if you want to add this.

In the end, I think that the resulting header is good. I didn't commit it because I don't know how you will redistribute it, but if you want to see... (github doesn't let me link .c files):
```c
#pragma once

/* Generated with cbindgen:0.24.6 */

#include <stdarg.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdlib.h>

/**
 * Result of CSS inlining operations
 */
typedef enum CssResult {
    /**
     * No error.
     */
    CSS_RESULT_OK,
    /**
     * Missing a stylesheet file.
     */
    CSS_RESULT_MISSING_STYLESHEET,
    /**
     * When loading a remote stylesheet, the file is not available.
     */
    CSS_RESULT_REMOTE_STYLESHEET_NOT_AVALABLE,
    /**
     * Error in the IO layer.
     */
    CSS_RESULT_IO_ERROR,
    /**
     * Error while parsing the CSS.
     */
    CSS_RESULT_INTERNAL_SELECTOR_PARSE_ERROR,
} CssResult;

/**
 * Configuration options for CSS inlining process.
 */
typedef struct CSSInlinerOptions {
    /**
     * Keep "style" tags after inlining.
     */
    bool keep_style_tags;
    /**
     * Keep "link" tags after inlining.
     */
    bool keep_link_tags;
    /**
     * Used for loading external stylesheets via relative URLs.
     */
    const char *base_url;
    /**
     * Size of base_url in bytes
     */
    size_t base_url_size;
    /**
     * Whether remote stylesheets should be loaded or not.
     */
    bool load_remote_stylesheets;
    /**
     * Additional CSS to inline.
     */
    const char *extra_css;
    /**
     * Size of extra_css in bytes
     */
    size_t extra_css_size;
    /**
     * Pre-allocate capacity for HTML nodes during parsing.
     * It can improve performance when you have an estimate of the number of nodes in your HTML document.
     */
    size_t preallocate_node_capacity;
} CSSInlinerOptions;

/**
 * @brief Inline CSS from @p input & write the result to @p output with @p options.
 * @param options configuration for the inliner.
 * @param input html to inline.
 * @param input_size size of @p input in bytes.
 * @param output buffer to save the inlined CSS.
 * @param output_size size of @p output in bytes.
 * @return a CSS_RESULT enum variant regarding if the operation was a success or an error ocurred
 */
enum CssResult inline_to(const struct CSSInlinerOptions *options,
                         const char *input,
                         size_t input_size,
                         char *output,
                         size_t output_size);

/**
 * @brief Creates an instance of CSSInlinerOptions with the default parameters.
 * @return a CSSInlinerOptions struct.
 */
struct CSSInlinerOptions css_inliner_default_options(void);
```

The command to do the generation is:
```shell
$ cbindgen --config cbindgen.toml --crate css-inline-c --lang c --output css_inline.h
```

laplus-sadness commented:
Forgot to add the cbindgen.toml. Will make another commit so wish me luck passing these 34 actions.

laplus-sadness commented:
I realized I wasn't checking for the end of the string, just getting the size for the user. I think reading until finding a '\0' is better honestly.

The header needs to be regenerated.

Stranger6667 commented:
Hi @laplus-sadness !

These are remarkable changes! I took a brief look and everything looks great on the high level :) I'll give a more detailed review during the weekend :) 

As for tests, I don't have a preference for C tests, so feel free to choose an option that is the most convenient for you :) Readme would also be great!

Stranger6667 commented on bindings/c/src/lib.rs:
```diff
@@ -0,0 +1,195 @@
+            Err(_) => return CssResult::InvalidOptions,
+        },
+    );
+    let html = str::from_utf8_unchecked(slice::from_raw_parts(
+        input as *const u8,
+        seek_for_string_end(input),
+    ));
```
I guess, we may use [CStr](https://doc.rust-lang.org/nightly/core/ffi/struct.CStr.html) here, which should do exactly this + will do UTF-8 validation (via `CStr::to_str`). UTF-8 will give some performance penalty (I expect it to be neglectable), but we'll avoid some possible UB.

Stranger6667 replied:
Or, maybe provide an “unchecked” version for those who don’t need such a check

laplus-sadness replied:
I changed the type to CStr, but it stills needs to be in UTF-8. I think the code is not less performant because checking is not that expensive. There are even [SIMD versions](https://lemire.me/blog/2020/10/20/ridiculously-fast-unicode-utf-8-validation/).

Stranger6667 replied:
Good point! Thanks! :)

Stranger6667 commented on bindings/c/tests/main.c:
```diff
@@ -0,0 +1,93 @@
+ * @param body the body to be used.
+ * @return true if MAX_SIZE is enough to fit the generated html, or else, false
+ */
+static bool make_hmtl(char *html, const char *style, const char *body) {
```
A typo - hmtl -> html

Stranger6667 commented on bindings/c/tests/main.c:
```diff
@@ -0,0 +1,93 @@
+  assert(strcmp(output, SAMPLE_INLINED) == 0);
+}
+
+static void test_output_size_too_smal(void) {
```
Typo `test_output_size_too_smal` -> `test_output_size_too_small`

Stranger6667 commented on bindings/c/src/lib.rs:
```diff
@@ -0,0 +1,202 @@
+        // Because write() attempts to write regardless of buf.len() being higher than self.size,
+        // we are going to check at each byte that we didn't cross the bounds
+        let mut bytes_written: usize = 0;
+        for byte in buf {
```
I am thinking about whether we can think about writing in chunks and if it would be too much of a complication and how much would it affect performance

laplus-sadness replied:
Good catch! I changed the code to use memcpy then.

Stranger6667 commented on bindings/c/src/lib.rs:
```diff
@@ -0,0 +1,202 @@
+    let mut buffer = CBuffer::new(output, output_size);
+    if let Err(e) = options.inline_to(html, &mut buffer) {
+        match e {
+            InlineError::IO(_) => return CssResult::IoError,
```
We can lift out `return` from here and place it before `match`

Stranger6667 commented on bindings/c/src/lib.rs:
```diff
@@ -0,0 +1,202 @@
+    pub keep_link_tags: bool,
+    /// Used for loading external stylesheets via relative URLs.
+    pub base_url: *const c_char,
+    /// Whether remote stylesheets should be loaded or not.
+    pub load_remote_stylesheets: bool,
```
As we use `repr(C)` would it be better to put `bool` a bit above? for the alignment reasons (the resulting `CssInlinerOptions` would be smaller)

laplus-sadness replied:
Didn't notice this.

Stranger6667 reviewed (changes requested):
Great work! I left a few minor comments! :) 

One of the important parts is also put things (clippy, tests, rustfmt, etc) into CI, similar to what we have for other bindings :)

I am not yet sure what would be the best way to distribute the bindings, but I'd be happy to discuss it :)

laplus-sadness commented:
Hello. I forgot to add this PR to draft because I would add more things later.

Will look into your reviews!

laplus-sadness commented:
Hello. I didn't write a README yet because we still need to discuss the best way to distribute the header and the library. But first, there are some changes I made:
- Renamed `inline_to()` to `css_iniline_to()` because it makes more sense considering that c has no namespaces.
- Changed `css_inline_to()`  to `#[must_use]` because it compiles with the `__attribute__((warn_unused_result))`, so when the caller forgets to deal with the return variable (the error) the compiler warns.
- Because you didn't specify a test framework, I settled with using just the `assert.h` header because it's simple and does the job. Problem is that the tests don't run in parallel and, if one fails, you can't see all the tests' results. But, because the asserts abort the program with a failure exit code, we can just compile the binary and run in a github workflow, so that's what I did.

Now, I think we should compile the library as dynamic instead of static because, if you would like to distribute it as a package in linux distributions later, it makes more sense. So, I think that bundling the `css_inline.h` header with the `libcss_inline.so` library in a github release would be a good way for now.

Stranger6667 commented:
Nice!

I like the recent changes, and I think that plain `assert` is fine for now :)

> Now, I think that we should compile the library as dynamic instead of static because, if you would like to distribute it as a package in linux distributions later, it makes more sense. So, I think that bundling the css_inline.h header with the libcss_inline.so library in a github release would be a good way for now.

Yes, I was thinking about Linux distros at some point and agree that a GitHub release bundled this way is a good starting point indeed.

Re: placing the header file. I guess, we can do some workarounds, i.e. hardcoding path in our CI build scripts or something like this, right?

Another thing I was thinking about is that it would probably be useful to have a CI job that checks before release whether the header file is up-to-date (assuming it will be committed to the repo). What do you think?

laplus-sadness commented:
Sorry for taking too long.

Because the header is dynamically generated, I don't know if we should commit it to the repo.

By following the rule that all the generated objects should live in the pipeline after we build them instead of adding them again after every change, I think that we should only get the header when we make a release.

Also, I made a readme based on the wasm binding. If you have some remarks, please say so!

Stranger6667 commented:
> By following the rule that all the generated objects should live in the pipeline after we build them instead of adding them again after every change, I think that we should only get the header when we make a release.

This is indeed consistent with what we already have, and frankly, anyone interested could just take a look at the releases page.

I checked the current state of PR - absolutely love it :) I think the only thing left is release automation, i.e. when a new tag is created, then we build the header and upload artifacts somewhere where they could be downloaded and then attached to a new release.

laplus-sadness commented:
@Stranger6667 I added a workflow that generates an artifact with the header and the library, but I don't know if it's working correctly since I have to add a new tag to test it.

Stranger6667 commented:
Amazing! Thank you so much! :) I’ll merge the PR and will work on the release for it
